From 8d8db65498ff390a4f0eb1f4ca08e25b734034a4 Mon Sep 17 00:00:00 2001
From: Ming Qian <ming.qian@nxp.com>
Date: Fri, 10 Jun 2022 15:31:45 +0800
Subject: [PATCH 58/91] MMFMWK-9186-2 video: Add support for Y312_LE

Y312_LE is a YUV format with 12-bits per component like YUV24,
expanded to 16bits.
Data in the 12 high bits, zeros in the 4 low bits,
arranged in little endian order.

Upstream-Status: Pending

Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/video/video-converter.c |  1 +
 gst-libs/gst/video/video-format.c    | 55 ++++++++++++++++++++++++++++
 gst-libs/gst/video/video-format.h    | 14 ++++++-
 gst-libs/gst/video/video-info.c      |  5 +++
 4 files changed, 73 insertions(+), 2 deletions(-)

diff --git a/gst-libs/gst/video/video-converter.c b/gst-libs/gst/video/video-converter.c
index 8f56c13..fe68a89 100644
--- a/gst-libs/gst/video/video-converter.c
+++ b/gst-libs/gst/video/video-converter.c
@@ -7368,6 +7368,7 @@ get_scale_format (GstVideoFormat format, gint plane)
     case GST_VIDEO_FORMAT_Y012_LE:
     case GST_VIDEO_FORMAT_Y212_BE:
     case GST_VIDEO_FORMAT_Y212_LE:
+    case GST_VIDEO_FORMAT_Y312_LE:
     case GST_VIDEO_FORMAT_Y412_BE:
     case GST_VIDEO_FORMAT_Y412_LE:
     case GST_VIDEO_FORMAT_NV12_8L128:
diff --git a/gst-libs/gst/video/video-format.c b/gst-libs/gst/video/video-format.c
index 47de5ce..cd42e12 100644
--- a/gst-libs/gst/video/video-format.c
+++ b/gst-libs/gst/video/video-format.c
@@ -6670,6 +6670,59 @@ pack_Y212_LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
   }
 }
 
+#define PACK_Y312_LE GST_VIDEO_FORMAT_AYUV64, unpack_Y312_LE, 1, pack_Y312_LE
+static void
+unpack_Y312_LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
+    gpointer dest, const gpointer data[GST_VIDEO_MAX_PLANES],
+    const gint stride[GST_VIDEO_MAX_PLANES], gint x, gint y, gint width)
+{
+  int i;
+  const guint16 *restrict s = GET_LINE (y);
+  guint16 *restrict d = dest;
+  guint16 Y, U, V;
+
+  s += x * 3;
+
+  for (i = 0; i < width; i++) {
+    Y = GST_READ_UINT16_LE (s + 3 * i + 0) & 0xfff0;
+    U = GST_READ_UINT16_LE (s + 3 * i + 1) & 0xfff0;
+    V = GST_READ_UINT16_LE (s + 3 * i + 2) & 0xfff0;
+
+    if (!(flags & GST_VIDEO_PACK_FLAG_TRUNCATE_RANGE)) {
+      U |= (U >> 12);
+      Y |= (Y >> 12);
+      V |= (V >> 12);
+    }
+
+    d[4 * i + 0] = 0xffff;
+    d[4 * i + 1] = Y;
+    d[4 * i + 2] = U;
+    d[4 * i + 3] = V;
+  }
+}
+
+static void
+pack_Y312_LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
+    const gpointer src, gint sstride, gpointer data[GST_VIDEO_MAX_PLANES],
+    const gint stride[GST_VIDEO_MAX_PLANES], GstVideoChromaSite chroma_site,
+    gint y, gint width)
+{
+  int i;
+  guint16 *restrict d = GET_LINE (y);
+  const guint16 *restrict s = src;
+  guint16 Y, U, V;
+
+  for (i = 0; i < width; i++) {
+    Y = s[4 * i + 1] & 0xfff0;
+    U = s[4 * i + 2] & 0xfff0;
+    V = s[4 * i + 3] & 0xfff0;
+
+    GST_WRITE_UINT16_LE (d + 3 * i + 0, Y);
+    GST_WRITE_UINT16_LE (d + 3 * i + 1, U);
+    GST_WRITE_UINT16_LE (d + 3 * i + 2, V);
+  }
+}
+
 #define PACK_Y412_BE GST_VIDEO_FORMAT_AYUV64, unpack_Y412_BE, 1, pack_Y412_BE
 static void
 unpack_Y412_BE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
@@ -7793,6 +7846,8 @@ static const VideoFormat formats[] = {
       SUB4444, PACK_RBGA),
   MAKE_GRAY_LE_FORMAT (Y012_LE, "raw video", DPTH12, PSTR2, PLANE0, OFFS0,
       SUB4, PACK_GRAY16_LE),
+  MAKE_YUV_LE_FORMAT (Y312_LE, "raw video", 0x00000000, DPTH12_12_12_HI,
+      PSTR222, PLANE0, OFFS0, SUB444, PACK_Y312_LE),
 };
 
 G_GNUC_END_IGNORE_DEPRECATIONS;
diff --git a/gst-libs/gst/video/video-format.h b/gst-libs/gst/video/video-format.h
index 361f628..804cccb 100644
--- a/gst-libs/gst/video/video-format.h
+++ b/gst-libs/gst/video/video-format.h
@@ -170,6 +170,7 @@ G_BEGIN_DECLS
  * @GST_VIDEO_FORMAT_GBR_16BE: planar 4:4:4 RGB, 16 bits per channel (Since: 1.24)
  * @GST_VIDEO_FORMAT_RBGA: packed RGB with alpha, 8 bits per channel (Since: 1.24)
  * @GST_VIDEO_FORMAT_Y012_LE: 12-bit grayscale, least significant byte first, expanded to 16bits, zeros in the 4 low bits,  (Since: 1.24)
+ * @GST_VIDEO_FORMAT_Y312_LE: packed 4:4:4 YUV, 12 bits per channel (Y-U-V) (Since: 1.24)
  *
  * Enum value describing the most common video formats.
  *
@@ -631,6 +632,15 @@ typedef enum {
    */
   GST_VIDEO_FORMAT_Y012_LE,
 
+  /**
+   * GST_VIDEO_FORMAT_Y312_LE:
+   *
+   * YUV 4:4:4 12 bits little endian.
+   * expanded to 16 bits, data in the 12 high bits, zeros in the 4 low bits
+   *
+   * Since: 1.24
+   */
+  GST_VIDEO_FORMAT_Y312_LE,
 } GstVideoFormat;
 
 #define GST_VIDEO_MAX_PLANES 4
@@ -1108,7 +1118,7 @@ gconstpointer  gst_video_format_get_palette          (GstVideoFormat format, gsi
     "RGB10A2_LE, A444, GBRA, AYUV, VUYA, RGBA, RBGA, ARGB, BGRA, ABGR, A422, " \
     "A420, AV12, Y444_16BE, GBR_16BE, Y444_16LE, GBR_16LE, v216, P016_BE, " \
     "P016_LE, Y444_12BE, GBR_12BE, Y444_12LE, GBR_12LE, I422_12BE, I422_12LE, " \
-    "Y212_BE, Y012_LE, Y212_LE, I420_12BE, I420_12LE, P012_BE, P012_LE, Y444_10BE, " \
+    "Y212_BE, Y012_LE, Y212_LE, Y312_LE, I420_12BE, I420_12LE, P012_BE, P012_LE, Y444_10BE, " \
     "GBR_10BE, Y444_10LE, GBR_10LE, r210, I422_10BE, I422_10LE, NV16_10LE32, " \
     "Y210, UYVP, v210, I420_10BE, I420_10LE, P010_10BE, MT2110R, MT2110T, " \
     "NV12_10BE_8L128, NV12_10LE40_4L4, P010_10LE, NV12_10LE40, NV12_10LE32, " \
@@ -1127,7 +1137,7 @@ gconstpointer  gst_video_format_get_palette          (GstVideoFormat format, gsi
     "A444, GBRA, AYUV, VUYA, RGBA, RBGA, ARGB, BGRA, ABGR, A422, A420, AV12, " \
     "Y444_16LE, GBR_16LE, Y444_16BE, GBR_16BE, v216, P016_LE, P016_BE, " \
     "Y444_12LE, GBR_12LE, Y444_12BE, GBR_12BE, I422_12LE, I422_12BE, Y012_LE, Y212_LE, " \
-    "Y212_BE, I420_12LE, I420_12BE, P012_LE, P012_BE, Y444_10LE, GBR_10LE, " \
+    "Y212_BE, Y312_LE, I420_12LE, I420_12BE, P012_LE, P012_BE, Y444_10LE, GBR_10LE, " \
     "Y444_10BE, GBR_10BE, r210, I422_10LE, I422_10BE, NV16_10LE32, Y210, UYVP, " \
     "v210, I420_10LE, I420_10BE, P010_10LE, NV12_10LE40, NV12_10LE32, " \
     "P010_10BE, MT2110R, MT2110T, NV12_10BE_8L128, NV12_10LE40_4L4, Y444, " \
diff --git a/gst-libs/gst/video/video-info.c b/gst-libs/gst/video/video-info.c
index 099e55c..c28d4a2 100644
--- a/gst-libs/gst/video/video-info.c
+++ b/gst-libs/gst/video/video-info.c
@@ -863,6 +863,11 @@ fill_planes (GstVideoInfo * info, gsize plane_size[GST_VIDEO_MAX_PLANES])
       info->offset[0] = 0;
       info->size = info->stride[0] * height;
       break;
+    case GST_VIDEO_FORMAT_Y312_LE:
+      info->stride[0] = width * 6;
+      info->offset[0] = 0;
+      info->size = info->stride[0] * height;
+      break;
     case GST_VIDEO_FORMAT_GRAY8:
       info->stride[0] = GST_ROUND_UP_4 (width);
       info->offset[0] = 0;
-- 
2.43.0

